zapcore/sampler.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapcore
+
+import (
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap/internal/hash"
+)
+
+const (
+	_numLevels        = maxLevel - minLevel
+	_countersPerLevel = 4096
+)
+
+func newCounters() *counters {
+	return &counters{}
+}
+
+type counters [_numLevels][_countersPerLevel]uint64
+
+func (c *counters) Inc(lvl Level, key string) uint64 {
+	i := lvl - minLevel
+	j := c.hash(key) % _countersPerLevel
+	return atomic.AddUint64(&c[i][j], 1)
+}
+
+func (c *counters) Reset(lvl Level, key string) {
+	i := lvl - minLevel
+	j := c.hash(key) % _countersPerLevel
+	atomic.StoreUint64(&c[i][j], 0)
+}
+
+func (c *counters) hash(key string) uint32 {
+	return hash.XSHRR(key) % _countersPerLevel
+}
+
+// Sample creates a facility that samples incoming entries.
+func Sample(fac Facility, tick time.Duration, first, thereafter int) Facility {
+	return &sampler{
+		Facility:   fac,
+		tick:       tick,
+		counts:     newCounters(),
+		first:      uint64(first),
+		thereafter: uint64(thereafter),
+	}
+}
+
+type sampler struct {
+	Facility
+
+	tick       time.Duration
+	counts     *counters
```

Comment:
We want the counts to be shared between parent and child loggers (so that sampling imposes a predictable limit on global logging load), so a pointer makes sense to me.